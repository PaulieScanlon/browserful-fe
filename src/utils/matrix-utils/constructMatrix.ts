import browserslist from 'browserslist';
import { browserDetails } from '../browserDetails';
import { fullArrays } from './full-arrays';

import { IVersion } from './types';

export const constructMatrix = (
  queryType: string,
  browserQuery: string,
  incQuery: Array<String>,
  excQuery: Array<String>
) => {
  const builtQuery = browserslist(`${browserQuery}`);

  const versionsList = fullArrays[queryType].map((browser: string) => {
    const [name, version] = browser.split(' ', 2);
    let concatQueryName = `${browserDetails[name].queryName} ${version}`;
    return {
      query: concatQueryName,
      name: name,
      version: version,
      isIncluded: builtQuery.includes(browser),
      hasOverride: incQuery.includes(concatQueryName)
        ? 'isIncluded'
        : false || excQuery.includes(concatQueryName)
        ? 'isExcluded'
        : false,
      platform: browserDetails[name].platform
    };
  });

  const namesList = versionsList.reduce((acc: Object, version: IVersion) => {
    const [name] = version.name.split(' ', 2);
    acc[name] = [].concat(acc[name] || [], {
      ...version
    });
    return {
      ...acc
    };
  }, {});

  const browserList = Object.keys(namesList).map(browser => {
    return {
      friendlyName: browserDetails[browser].friendlyName,
      logo: browserDetails[browser].logo,
      platform: browserDetails[browser].platform,
      totalCount: namesList[browser] ? namesList[browser].length : 0,
      includedCount: namesList[browser]
        ? namesList[browser].filter(version => version.isIncluded === true)
            .length
        : 0,
      expandCard: true,
      versions: namesList[browser]
    };
  });

  return {
    browserList: browserList
  };
};
