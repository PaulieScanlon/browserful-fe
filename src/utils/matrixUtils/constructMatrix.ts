import browserslist from 'browserslist';
import { browserDetails } from './browserDetails';
import { platform, overrides } from './enums';

export const constructMatrix = (
  browserQuery: string,
  comparisonQuery: string,
  incQuery: Array<String>,
  excQuery: Array<String>
) => {
  const builtQuery = browserslist(`${browserQuery}`);
  const builtComparison = browserslist(`${comparisonQuery}`);

  const getIsIncluded = version => version.isIncluded;
  const getIsExcluded = version => !version.isIncluded;

  const getOverrride = q =>
    incQuery.includes(q)
      ? overrides.IS_INCLUDED
      : false || excQuery.includes(q)
      ? overrides.IS_EXCLUDED
      : false;

  const getTotalIncluded = (sum, item) => sum + item.totalIncluded;
  const getTotalExcluded = (sum, item) => sum + item.totalExcluded;

  const getTotal = (sum, item) => sum + item.total;

  const browserObject: any = builtComparison
    .map(result => {
      const [name, version] = result.split(' ', 2);
      return {
        name: name,
        version: version,
        platform: browserDetails[name].platform
      };
    })
    .reduce((browser, item) => {
      const { name, version } = item;

      browser[name] = browser[name] || {
        ...browserDetails[name],
        totalIncluded: 0,
        totalExcluded: 0,
        total: 0,
        expandCard: true,
        versions: []
      };

      let q = `${name} ${version}`;

      browser[name].versions.push({
        friendlyName: browserDetails[name].friendlyName,
        query: q,
        version: version,
        isIncluded:
          getOverrride(q) === overrides.IS_EXCLUDED
            ? false
            : builtQuery.includes(q),
        hasOverride: getOverrride(q),
        platform: browserDetails[name].platform
      });

      browser[name].totalIncluded = browser[name].versions.filter(
        getIsIncluded
      ).length;

      browser[name].totalExcluded = browser[name].versions.filter(
        getIsExcluded
      ).length;

      browser[name].total = browser[name].versions.length
        ? browser[name].versions.length
        : 0;

      return browser;
    }, {});

  const platformObject = Object.keys(browserObject)
    .map(browser => {
      const [name] = browser.split(' ', 1);
      return browserObject[name];
    })
    .reduce(
      (platforms, item) => {
        const { platform } = item;

        platforms[platform].push(item);

        return platforms;
      },
      { [platform.DESKTOP]: [], [platform.MOBILE]: [] }
    );

  return {
    browserList: {
      desktop: platformObject.desktop,
      mobile: platformObject.mobile
    },
    includedTotal: {
      desktop: platformObject.desktop.reduce(getTotalIncluded, 0),
      mobile: platformObject.mobile.reduce(getTotalIncluded, 0)
    },
    excludedTotal: {
      desktop: platformObject.desktop.reduce(getTotalExcluded, 0),
      mobile: platformObject.mobile.reduce(getTotalExcluded, 0)
    },
    total: {
      desktop: platformObject.desktop.reduce(getTotal, 0),
      mobile: platformObject.mobile.reduce(getTotal, 0)
    }
  };
};
