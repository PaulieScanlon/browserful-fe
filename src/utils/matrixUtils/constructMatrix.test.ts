import browserslist from 'browserslist';
import { constructMatrix } from './constructMatrix';

describe('constructMatrix', () => {
  it('matches snapshot', () => {
    const comparisonQuery = browserslist('chrome 70, chromeandroid 69');

    const matrix = constructMatrix(
      'chrome 70, chromeandroid 69',
      comparisonQuery,
      [],
      []
    );

    // console.log(JSON.stringify(matrix, null, 2));
    expect(matrix).toMatchSnapshot();
  });

  it('returns correct keys for browserList', () => {
    const comparisonQuery = browserslist('chrome 70, chromeandroid 69');

    const matrix = constructMatrix(
      'chrome 70, chromeandroid 69',
      comparisonQuery,
      [],
      []
    );

    expect(matrix.browserList).toBeDefined();
    expect(matrix.browserList.desktop).toBeDefined();
    expect(matrix.browserList.mobile).toBeDefined();
  });

  it('returns correct shape for browserObject', () => {
    const comparisonQuery = browserslist('chrome 70');

    const matrix = constructMatrix('chrome 70', comparisonQuery, [], []);

    const {
      friendlyName,
      logo,
      platform,
      totalIncluded,
      totalExcluded,
      total,
      expandCard
    } = matrix.browserList.desktop[0];

    expect(typeof friendlyName).toEqual('string');
    expect(typeof logo).toEqual('string');
    expect(typeof platform).toEqual('string');
    expect(typeof totalIncluded).toEqual('number');
    expect(typeof totalExcluded).toEqual('number');
    expect(typeof total).toEqual('number');
    expect(typeof expandCard).toEqual('boolean');
  });

  it('returns correct shape for browserObject.versions', () => {
    const comparisonQuery = browserslist('chrome 70');

    const matrix = constructMatrix('chrome 70', comparisonQuery, [], []);

    const {
      friendlyName,
      query,
      version,
      isIncluded,
      hasOverride,
      platform
    } = matrix.browserList.desktop[0].versions[0];

    expect(typeof friendlyName).toEqual('string');
    expect(typeof query).toEqual('string');
    expect(typeof version).toEqual('string');
    expect(typeof isIncluded).toEqual('boolean');
    expect(typeof hasOverride).toEqual('boolean');
    expect(typeof platform).toEqual('string');
  });

  it('returns correct values for includedList', () => {
    const comparisonQuery = browserslist('chrome 70, chromeandroid 69');

    const matrix = constructMatrix(
      'chrome 70, chromeandroid 69',
      comparisonQuery,
      [],
      []
    );

    expect(matrix.includedList).toEqual(
      expect.objectContaining({
        desktop: 1,
        mobile: 1
      })
    );
  });

  it('returns correct values for excludedList', () => {
    const comparisonQuery = browserslist('chrome 70, chrome 69');

    const matrix = constructMatrix(
      'chrome 70, chromeandroid 69',
      comparisonQuery,
      [],
      []
    );

    expect(matrix.excludedList).toEqual(
      expect.objectContaining({
        desktop: 1,
        mobile: 0
      })
    );
  });

  it('returns correct values for total', () => {
    const comparisonQuery = browserslist(
      'chrome 70, chrome 69, chrome 68, chrome 67'
    );

    const matrix = constructMatrix('chrome 70', comparisonQuery, [], []);

    expect(matrix.total).toEqual(
      expect.objectContaining({
        desktop: 4,
        mobile: 0
      })
    );
  });

  it('is correctly isIncluded if isIncluded ovverride is set', () => {
    const comparisonQuery = browserslist('chrome 70');

    const matrix = constructMatrix(
      'chrome 70',
      comparisonQuery,
      ['chrome 70'],
      []
    );

    const {
      hasOverride,
      isIncluded
    } = matrix.browserList.desktop[0].versions[0];

    expect(isIncluded).toBe(true);
    expect(hasOverride).toEqual('isIncluded');
    expect(typeof hasOverride).toEqual('string');
  });

  it('is correctly isIncluded if isExcluded ovverride is set', () => {
    const comparisonQuery = browserslist('chrome 70');

    const matrix = constructMatrix(
      'chrome 70',
      comparisonQuery,
      [],
      ['chrome 70']
    );

    const {
      hasOverride,
      isIncluded
    } = matrix.browserList.desktop[0].versions[0];

    expect(isIncluded).toBe(false);
    expect(hasOverride).toEqual('isExcluded');
    expect(typeof hasOverride).toEqual('string');
  });
});
