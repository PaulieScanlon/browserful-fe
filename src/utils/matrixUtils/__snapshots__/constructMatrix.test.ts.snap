// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructMatrix matches snapshot 1`] = `
Object {
  "browserList": Object {
    "desktop": Array [
      Object {
        "expandCard": true,
        "friendlyName": "Chrome",
        "logo": "chrome",
        "platform": "desktop",
        "queryName": "chrome",
        "total": 1,
        "totalExcluded": 0,
        "totalIncluded": 1,
        "versions": Array [
          Object {
            "friendlyName": "Chrome",
            "hasOverride": false,
            "isIncluded": true,
            "platform": "desktop",
            "query": "chrome 70",
            "version": "70",
          },
        ],
      },
    ],
    "mobile": Array [
      Object {
        "expandCard": true,
        "friendlyName": "Chrome (Android)",
        "logo": "chrome-for-android",
        "platform": "mobile",
        "queryName": "chromeandroid",
        "total": 1,
        "totalExcluded": 0,
        "totalIncluded": 1,
        "versions": Array [
          Object {
            "friendlyName": "Chrome (Android)",
            "hasOverride": false,
            "isIncluded": true,
            "platform": "mobile",
            "query": "and_chr 73",
            "version": "73",
          },
        ],
      },
    ],
  },
  "excludedTotal": Object {
    "desktop": 0,
    "mobile": 0,
  },
  "includedTotal": Object {
    "desktop": 1,
    "mobile": 1,
  },
  "total": Object {
    "desktop": 1,
    "mobile": 1,
  },
}
`;
